### ISSUE TITLE
As a user, when I include both features and a feature_service group in an evaluation for only bias fraction and again for just POD, I expect outputs for the same features and thresholds

### ISSUE THREAD

**Issue by HankHerr-NOAA** [NOAA-OWP/wres#319]
See VLab User Support ticket #134706. The name of this ticket is nearly identical to that one, but feel free to rename it later if there is a better, more searchable name to use.

For a given evaluation that includes both `features` and a `feature_service` `group`, with the declaration in that ticket (comment 1), if I run it with POD and FAR metrics, I get results for both sets of features, but, if I specify bias fraction or sample size, even if POD and FAR are still included, I only get results for the `features`, not the `group`. A log message indicates that the features in the `group` were dropped due to lack of thresholds; for example:

> 2024-09-13T18:46:34.621+0000 WARN DeclarationUtilities Discovered 11 feature(s) for which thresholds were not available. These features have been removed from the evaluation. The features are: [{"left":{"name":"04195500"},"right":{"name":"WDVO1"}}, {"left":{"name":"04199000"},"right":{"name":"MILO1"}}, {"left":{"name":"04208000"},"right":{"name":"INDO1"}}, {"left":{"name":"04206000"},"right":{"name":"OLPO1"}}, {"left":{"name":"04197137"},"right":{"name":"TFFO1"}}, {"left":{"name":"04196500"},"right":{"name":"UPPO1"}}, {"left":{"name":"04209000"},"right":{"name":"WBYO1"}}, {"left":{"name":"04200500"},"right":{"name":"ELRO1"}}, {"left":{"name":"04198000"},"right":{"name":"FFMO1"}}, {"left":{"name":"04199500"},"right":{"name":"VERO1"}}, {"left":{"name":"04212100"},"right":{"name":"PNVO1"}}].

That log message is not output for the POD/FAR only run.  

I'll upload the data needed to reproduce and the declaration in comment one. Thanks,

Hank

**Comment by HankHerr-NOAA** [comment]
A declaration based on the original declaration that reproduces the issue with a single `features` entry for MEDP1 is provided below. Links to thresholds and RFC forecast data to support that declaration are then provided. Thanks,

Hank

==========

```yaml
label: CLE - REPRODUCE
observed:
  label: USGS NWIS Stage Observations
  sources:
    uri: https://nwis.waterservices.usgs.gov/nwis/iv
    interface: usgs nwis
  variable: "00065" 
  type: observations
predicted:
  label: WRDS AHPS API Stage Forecasts
  sources:
    uri: https://[WRDS]/api/rfc_forecast/v2.0/forecast/stage
    interface: wrds ahps
  variable:
    name: HG
    label: stage
    aliases:
      - HP
      - HT
  type: single valued forecasts
features:
  - {observed: "03023100", predicted: MEDP1}
#  - {observed: "03094000", predicted: LEAO1}
#  - {observed: "411402080492400", predicted: WRNO1}
#  - {observed: "03098600", predicted: YGWO1}
#  - {observed: "03117000", predicted: MLNO1}
#  - {observed: "03118500", predicted: NHSO1}
#  - {observed: "03136500", predicted: MVEO1}
#  - {observed: "03139000", predicted: KILO1}
#  - {observed: "03217500", predicted: LARO1}
#  - {observed: "03219500", predicted: PRGO1}
#  - {observed: "04189000", predicted: FDYO1}
#  - {observed: "04193500", predicted: WTVO1}
feature_service:
  uri: https://[WRDS]/api/location/v3.0/metadata
  group: tag
  value: OHRFC_SGTL
unit: FT
reference_dates:
  minimum: 2024-09-01T00:00:00Z
  maximum: 2024-09-13T00:00:00Z
valid_dates:
  minimum: 2024-09-01T00:00:00Z
  maximum: 2024-09-13T00:00:00Z
lead_times:
  minimum: 0
  maximum: 120
  unit: hours
threshold_sources:
  uri: https://[WRDS]/api/location/v3.0/nws_threshold/
  operator: greater equal
  feature_name_from: predicted
  parameter: stage
  provider: NWS-NRLDB
metrics:
  - probability of detection
  - false alarm ratio
duration_format: days
output_formats:
  - csv2
  - png
  - pairs
 ```

[134706_thresholds_MEDP1.json](https://github.com/user-attachments/files/17061661/134706_thresholds_MEDP1.json)
[134706_thresholds_OHRFC_SGTL.json](https://github.com/user-attachments/files/17061663/134706_thresholds_OHRFC_SGTL.json)
[134706_feature_group_tag_OHRFC_SGTL_stage.json](https://github.com/user-attachments/files/17061667/134706_feature_group_tag_OHRFC_SGTL_stage.json)
[134706_rfc_forecast_medp1_stage.json](https://github.com/user-attachments/files/17061668/134706_rfc_forecast_medp1_stage.json)

**Comment by james-d-brown** [comment]
Managed to reproduce locally.

The issue appears to be with the behavior of `DeclarationUtilities::removeFeaturesWithoutThresholds`, which attempts to remove from the evaluation any geographic features that have no corresponding thresholds available.

However, the involvement of the `metrics` declared is a bit bizarre and will need to be explained.

**Comment by james-d-brown** [comment]
> However, the involvement of the `metrics` declared is a bit bizarre and will need to be explained.

The superficial explanation is as follows. Since the `bias fraction` is a continuous metric, there is an `all data` threshold by default. There is no such default when declaring only categorical metrics, like the `probability of detection` and `false alarm ratio`. In this situation, there is an `all data` threshold associated with the `observed` feature name of `03023100` (`predicted` name = `MEDP1`) when declaring at least one continuous metric, otherwise not. This changes the behavior of the method in question that filters out features without thresholds.

Bottom line, that method needs to be cornered with a failing unit test and then fixed.

**Comment by james-d-brown** [comment]
From a user's POV, the key to the difference in behavior is alluded to by the presence or absence of the following warning. This warning ONLY appears when declaring continuous metrics alone, and it indicates that most of the features have been filtered away. So there is a hook in the log about the origin of the bug, but it is a bug nonetheless.

```
2024-09-19T15:26:12.576+0000 23924 [main] WARN DeclarationUtilities - Discovered 11 feature(s) for which thresholds were not available. These features have been removed from the evaluation. The features are: [{"left":{"name":"04212100"},"right":{"name":"PNVO1"}}, {"left":{"name":"04199500"},"right":{"name":"VERO1"}}, {"left":{"name":"04208000"},"right":{"name":"INDO1"}}, {"left":{"name":"04195500"},"right":{"name":"WDVO1"}}, {"left":{"name":"04199000"},"right":{"name":"MILO1"}}, {"left":{"name":"04209000"},"right":{"name":"WBYO1"}}, {"left":{"name":"04197137"},"right":{"name":"TFFO1"}}, {"left":{"name":"04206000"},"right":{"name":"OLPO1"}}, {"left":{"name":"04198000"},"right":{"name":"FFMO1"}}, {"left":{"name":"04196500"},"right":{"name":"UPPO1"}}, {"left":{"name":"04200500"},"right":{"name":"ELRO1"}}].
```

**Comment by james-d-brown** [comment]
( I am largely repeating the OP here, but just working through things systematically and confirming the OP. )

**Comment by james-d-brown** [comment]
Have the unit test passing. Will push momentarily.

**Comment by james-d-brown** [comment]
Pushed the fix for the OP. Confirmed that the evaluation in the OP now produces statistics for all features when only declaring the `bias fraction` (or any other continuous metrics alone).

**Comment by james-d-brown** [comment]
User reported a second issue in the same support ticket as follows:

```
Thank you for all your work digging into this strange issue! I wanted to add one more piece of info I learned this morning:

I tried running a bias fraction evaluation for the group (OHRFC_SGTL) on its own, with thresholds, and while the standard out did specify that it found 11 features with thresholds, it still returned a single bias fraction value; the CSV output likewise did not even include a column for thresholds. However, when I tried the same bias fraction evaluation with thresholds for the individual 11 features (no group, just listing out all of the features in OHRFC_SGTL), it worked perfectly. The output included bias fraction for all thresholds, had threshold columns in the results, etc. Very strange.
```

I am pretty sure this is a separate issue, but I will attempt to reproduce, debug and address here too.

**Comment by james-d-brown** [comment]
Managed to reproduce the issue by adding a low threshold to TFFO1 and then witnessing statistics for low flows when executed within a feature service group, else for all data only when executed as an explicitly declared feature.

**Comment by james-d-brown** [comment]
Issue appears to be with the interpolation of thresholds. In this context, interpolation means "rationalizing" the declaration to use hints in various places to fill in gaps. One of these gaps is the metric-specific thresholds. This interpolation step involves taking thresholds declared or ingested in any context and adding the equivalent declaration of thresholds per metric and feature, which is the canonical declaration that drives subsequent steps. It appears this interpolation step is only interpolating the `all data` thresholds to the individual metrics when the thresholds originate from the wrds source, and not the other thresholds.

**Comment by james-d-brown** [comment]
In other words, an issue with `DeclarationInterpolator::interpolate` (rather, one of the individual interpolation steps performed by that method) for which a failing unit test is needed to expose the behavior.

**Comment by james-d-brown** [comment]
Problem is here:

```
            // Special case: an 'all data' threshold is declared explicitly for a metric. This indicates that the metric
            // should be computed for 'all data' only, as the 'all data' threshold is otherwise added by default, which
            // would render the request meaningless
            if ( parametersBuilder.thresholds()
                                  .size() == 1
                 && parametersBuilder.thresholds()
                                     .iterator()
                                     .next()
                                     .threshold()
                                     .equals( DeclarationUtilities.ALL_DATA_THRESHOLD.threshold() ) )
            {
                LOGGER.debug( "The metric {} contained a threshold override for 'all data' only.", name );
                globalThresholds = Map.of();
            }
```

In this case, there is no override for all data, but an all data threshold had already been interpolated and was interpreted as an override, so no other thresholds were interpolated. That's the problem. Fixing will be trickier because this feature was added intentionally for the situation where a particular metric  (e.g. a HEFS evaluation with a QQ diagram) required the `all data` threshold only, despite a larger set of global thresholds. It is probably a sequencing thing, not interpolating the all data threshold until after other threshold interpolation has been performed.

**Comment by james-d-brown** [comment]
Done and ready for UAT in due course.

**Comment by epag** [comment]
UAT passes in job 3587499138796272709
